cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,283 @@
+	if len(bucketArr) < 2 {
+		return "", errors.New("parsing bucket")
+	}
+	prefix := strings.TrimLeft(c.BackupDest, bucketArr[0]+"/")
+	bucket := bucketArr[0]
+	s3, err := storage.NewS3(strings.TrimPrefix(strings.TrimPrefix(c.Endpoint, "https://"), "http://"), c.AccessKeyID, c.AccessKey, bucket, c.Region, strings.HasPrefix(c.Endpoint, "https"))
+	if err != nil {
+		return "", errors.Wrap(err, "new storage manager")
+	}
+
+	infoObj, err := s3.GetObject(prefix + "/xtrabackup_info.00000000000000000000") //TODO: work with compressed file
+	if err != nil {
+		return "", errors.Wrap(err, "get object")
+	}
+
+	lastGTID, err := getLastBackupGTID(infoObj)
+	if err != nil {
+		return "", errors.Wrap(err, "get last backup gtid")
+	}
+	return lastGTID, nil
+}
+
+const (
+	Latest      RecoverType = "latest"      // recover to the latest existing binlog
+	Date        RecoverType = "date"        // recover to exact date
+	Transaction RecoverType = "transaction" // recover to needed trunsaction
+	Skip        RecoverType = "skip"        // skip transactions
+)
+
+func (r *Recoverer) Run() error {
+	host, err := network.GetPXCLastHost(r.pxcServiceName)
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, r.pxcUser, r.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	r.db = pxc
+
+	err = r.setBinlogs()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	err = r.recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+
+	return nil
+}
+
+func (r *Recoverer) recover() error {
+	host, err := network.GetPXCLastHost(r.pxcServiceName)
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+
+	flags := ""
+	endTime := time.Time{}
+	// TODO: add logic for all types
+	switch r.recoverType {
+	case Skip:
+		flags = " --exclude-gtids=" + r.gtidSet
+	case Transaction:
+	case Date:
+		flags = ` --stop-datetime="` + r.recoverTime + `"`
+
+		format := "2006-01-02 15:04:05"
+		t, err := time.Parse(format, r.recoverTime)
+		if err != nil {
+			return errors.Wrap(err, "parse date")
+		}
+		endTime = t
+	case Latest:
+	default:
+		return errors.New("wrong recover type")
+	}
+
+	for _, binlog := range r.binlogs {
+		log.Println("working with", binlog)
+
+		if r.recoverType == Date {
+			binlogArr := strings.Split(binlog, "_")
+			if len(binlogArr) < 2 {
+				return errors.New("get timestamp from binlog name")
+			}
+			binlogTime, err := strconv.ParseInt(binlogArr[1], 10, 64)
+			if err != nil {
+				return errors.Wrap(err, "get binlog time")
+			}
+			if binlogTime > endTime.Unix() {
+				return nil
+			}
+		}
+
+		binlogObj, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get obj")
+		}
+
+		err = os.Setenv("MYSQL_PWD", os.Getenv("PXC_PASS"))
+		if err != nil {
+			return errors.Wrap(err, "set mysql pwd env var")
+		}
+
+		cmdString := "cat | mysqlbinlog" + flags + " - | mysql -h" + host + " -u" + r.pxcUser
```

Comment:
I don't think we need `cat |` here

Reply:
removed https://github.com/percona/percona-xtradb-cluster-operator/pull/673/commits/059533512393f8112ca47ab30263ab832efbaef8